### PR TITLE
Mark OPENSSL_armcap_P .hidden in arm asm

### PR DESCRIPTION
**Pull request by dg0yt**
Fixes https://github.com/openssl/openssl/pull/21583#issuecomment-1727057735:
~~~
ld: error: relocation R_ARM_REL32 cannot be used against symbol 'OPENSSL_armcap_P'; recompile with -fPIC
~~~
when linking a static build of openssl to a shared library for 32 bit arm on Android,
since https://github.com/openssl/openssl/pull/21583.

This source change has one observable change in readelf output (e.g. for `crypto/sha/libcrypto-lib-sha512-armv4.o`):
~~~diff
 Symbol table '.symtab' contains 12 entries:
    Num:    Value  Size Type    Bind   Vis      Ndx Name
      0: 00000000     0 NOTYPE  LOCAL  DEFAULT  UND 
      1: 00000000     0 SECTION LOCAL  DEFAULT    2 
      2: 00000000   640 OBJECT  LOCAL  DEFAULT    2 K512
      3: 00000000     0 NOTYPE  LOCAL  DEFAULT    2 $d.0
      4: 000002a0     0 NOTYPE  LOCAL  DEFAULT    2 $t.1
      5: 0000183a     0 NOTYPE  LOCAL  DEFAULT    2 $d.2
      6: 00000000     0 SECTION LOCAL  DEFAULT    6 
      7: 00000000     0 SECTION LOCAL  DEFAULT    8 
      8: 00000000     0 SECTION LOCAL  DEFAULT   11 
-     9: 00000000     0 NOTYPE  GLOBAL DEFAULT  UND OPENSSL_armcap_P
+     9: 00000000     0 NOTYPE  GLOBAL HIDDEN   UND OPENSSL_armcap_P
     10: 000002a1  1134 FUNC    GLOBAL DEFAULT    2 sha512_block_data_order
     11: 00000711  4394 FUNC    GLOBAL DEFAULT    2 sha512_block_data_order_n
 
~~~

This change was automated using
~~~
sed -e '/[.]hidden.*OPENSSL_armcap_P/d; /[.]extern.*OPENSSL_armcap_P/ {p; s/extern/hidden/ }' -i -- crypto/*arm*pl crypto/*/asm/*arm*pl
~~~
with the filepath patterns derived from https://github.com/openssl/openssl/pull/21583.

### CODE REVIEW

**Comment by dg0yt** [comment]
More reading: 
Commit message in https://github.com/google/boringssl/commit/16e38b2b8f50a3d048f61d2979d5ceddacd70fc3. 
https://github.com/android/ndk/issues/191.

File: crypto/armv4cpuid.pl
Comment (tom-cosgrove-arm) [inline crypto/armv4cpuid.pl:296]: Rather than making an identical change in every Perl-asm file, and then requiring this same pairing in every new Perl-asm that gets added in the future, wouldn't it be better to have `arm-xlate.pl` apply this directly? (At least, that's the way I've been looking at fixing this)

Comment (dg0yt) [reply crypto/armv4cpuid.pl:296]: I can't say much about openssl's perl asm stuff. 
First, it should be validated that `.hidden` is the right thing here.

Comment (tom-cosgrove-arm) [reply crypto/armv4cpuid.pl:296]: Note that this isn't what BoringSSL has anyway - they still have `.comm` rather than `.extern`. One of our recent-ish changes made that here, but I am now revisiting whether that was the best thing to do, as `arm-xlate.pl` actually removes the `.extern` altogether, but `.comm` keeps it in

**Comment by kroeckx** [comment]
The .comm vs .extern is not relevant. The reason the .hidden solves the problem is that it now can not be a symbol overridden by another library or executable. There is no reason to support that.

**Comment by talregev** [comment]
@t8m 
Any news?

**Comment by tom-cosgrove-arm** [comment]
I am looking at this (slowly, sorry)

**Comment by BillyONeal** [comment]
@tom-cosgrove-arm @t8m Given those tags just added does that mean this solution is acceptable to the OpenSSL maintainers? (I'm trying to confirm that it's safe to merge @dg0yt 's patch of this in vcpkg)

**Comment by tom-cosgrove-arm** [comment]
@BillyONeal I don't think the final fix will look like this, but not sure exactly what it should be at the moment. This is now top of my OpenSSL list, though

**Comment by BillyONeal** [comment]
@tom-cosgrove-arm OK, thanks, I'll drop this from daily checks. Please poke me when this is fixed if you remember :)

**Comment by kroeckx** [comment]
I see no problem with the patch as is

**Comment by BillyONeal** [comment]
> I see no problem with the patch as is

As in, it passes for the platforms vcpkg ships to but there are others that openssl cares about perl scripts or something like that, so I should merge @dg0yt's PR?

**Comment by tom-cosgrove-arm** [comment]
As it stands, this patch would require us to add `.hidden` every time we add a new assembler file - specifically, to remember to do that. That's not good for maintainability - it will just cause this issue to recur every now and them. Moreover, I don't understand why this patch works - declaring the variable `.extern` should be sufficient. However, the xlate script removes `.extern` for unknown reasons, and that doesn't seem right either. Hence: doing more investigations to understand the root cause, rather than adding a workaround

**Comment by talregev** [comment]
@tom-cosgrove-arm We just want to know if the current change is safe to merge in vcpkg side. meaning everytime that someone will use vcpkg openssl with this patch it will work correctly and it will be safe to use.
When you find long term solution and it will be in different PR or / and it will be on the next version, when vcpkg will update the openssl next version it will take the long term solution with the new version / new commit.

Thank you for your time.

**Comment by dg0yt** [comment]
> However, the xlate script removes `.extern` for unknown reasons, and that doesn't seem right either.

Can you verify this? I don't see in the script, and I don't see the effect in the build dir. E.g.:
~~~
buildtrees/openssl/arm-neon-android-dbg/crypto/sha/sha512-armv4.S:.extern       OPENSSL_armcap_P
buildtrees/openssl/arm-neon-android-dbg/crypto/sha/sha512-armv4.S:.hidden       OPENSSL_armcap_P
~~~
And IIUC, OPENSSL_armcap_P would be undeclared without `.extern`.

**Comment by kroeckx** [comment]
As the linker complains, the code is not position independent code (PIC). With a symbol that's exported, the dynamic linker needs to be able to support replacing the symbol by one from an other library (or executable). To be able to do that, the code needs to be PIC, and you'll get a different relocation type. So there are a few solutions:
- Don't export the symbol. This can be done by marking it hidden. The linker will then know no other library can replace it, and it doesn't need to be PIC.
- Rewrite the code to be PIC.
- Tell the linker it doesn't need to support replacing the symbol, by building the shared library using -Bsymbolic.

**Comment by kroeckx** [comment]
Note that OpenSSL builds using -Bsymbolic by default, which is why it doesn't show up for most people.

**Comment by t8m** [comment]
The problem is now independently reported in #22414

**Comment by Eimji** [comment]
Hello

I tried the fix here, it works, I can compile openssl for my Android application.

Thanks.

**Comment by dg0yt** [comment]
Is the fix to a present problem still blocked by concerns about files which might be added in the future?
Isn't it possible to verify whether "the xlate script removes `.extern`"?

**Comment by dg0yt** [comment]
Still waiting.

**Comment by dg0yt** [comment]
@talregev Please don't put oil on the fire.

**Comment by tom-cosgrove-arm** [comment]
Yes, still looking at it. The issue recently fixed by https://github.com/openssl/openssl/pull/22880 meant that some of the changes we made to the handling of `OPENSSL_armcap_P` could be rolled back, so I am looking at doing that as an option

**Comment by dg0yt** [comment]
Another six weeks. Why contribute at all.

**Comment by t8m** [comment]
@tom-cosgrove-arm any progress on this?

**Comment by ospfranco** [comment]
Facing this issue. Seems there is no possible workaround at the moment. Is there anything that can be done to merge this faster? Is sponsoring possible?

https://github.com/android/ndk/issues/1992

**Comment by dg0yt** [comment]
Spring 2024.

**Comment by t8m** [comment]
@tom-cosgrove-arm ping?

**Comment by tom-cosgrove-arm** [comment]
Yes, sorry, won't be this week, but still on my radar

**Comment by dg0yt** [comment]
May 2024.

**Comment by BillyONeal** [comment]
June :)

**Comment by BillyONeal** [comment]
> @BillyONeal I don't think the final fix will look like this, but not sure exactly what it should be at the moment. This is now top of my OpenSSL list, though
> 
> (The reason I don't think the final fix will look like this is at a minimum the duplication, and I think the fix will need something in `arm-xlate.pl` too)

Would you be angry at me / vcpkg for merging this as a patch in vcpkg ( https://github.com/microsoft/vcpkg/pull/33955 )? I understand that you might not want this exact change yourselves due to duplicated edits but I've interpreted your statement above as an "explicit no from upstream".

Thanks!

**Comment by tom-cosgrove-arm** [comment]
> Would you be angry at me / vcpkg for merging this as a patch in vcpkg ( [microsoft/vcpkg#33955](https://github.com/microsoft/vcpkg/pull/33955) )? I understand that you might not want this exact change yourselves due to duplicated edits but I've interpreted your statement above as an "explicit no from upstream".

Given how long it's taking for me to get back to this, no, of course I wouldn't be angry! This is still on my list, and I'd like to get a proper fix in, but do feel free to put a workaround patch in vcpkg

**Comment by dg0yt** [comment]
One year.

**Comment by ospfranco** [comment]
Here is a tutorial for Rust on how to patch and manually compile the latest versions of OpenSSL for Android

https://ospfranco.com/compiling-openssl-in-rust-for-android/

But yeah, still waiting for a fix here

**Review by tom-cosgrove-arm — APPROVED** [review]
Okay, after watching the fun with `OPENSSL_ia32cap_P`, and that this has been raised again with #26001, I think this really is going to be the best solution, rather than doing anything with `arm-xlate.pl` that may come back to bite us later (e.g. the fact that it silently removes all `.extern`s, which I learnt from the `OPENSSL_ia32cap_P` work is ignored by `gas`.

**Comment by tom-cosgrove-arm** [comment]
This should fix #25601 and #22414 and close #26001

**Review by t8m — APPROVED** [review]

**Review by t-j-h — APPROVED** [review]

**Comment by t-j-h** [comment]
I'm also okay with this being backfitted to active branches.

**Comment by tom-cosgrove-arm** [comment]
CI is now passing

**Comment by t8m** [comment]
Merged to the master, 3.4, 3.3, 3.2 and 3.1 branches. Thank you for your contribution.

**Comment by epoberezkin** [comment]
@t8m could it be added to 3.0.x? we have some weird armv7a issues since switching to 3.0.15

**Pull request closed without merging**
